src/xform_to_datalog/datalog_facts.cc

```diff
@@ -0,0 +1,149 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#include "absl/container/flat_hash_map.h"
+#include "src/xform_to_datalog/datalog_facts.h"
+
+#include "src/xform_to_datalog/arcs_manifest_tree/particle_spec.h"
+
+namespace raksha::xform_to_datalog {
+
+namespace ir = raksha::ir;
+namespace types = raksha::ir::types;
+namespace amt = raksha::xform_to_datalog::arcs_manifest_tree;
+
+// Traverse the substructures of the manifest proto to create datalog fact
+// objects.
+// #TODO(#107): In the interest of prototyping speed, I wrote out the
+//  entirety of the traversal of instantiation structures as one big
+//  traversal. This works, but it is hard to read and hard to test. We should
+//  break this up.
+DatalogFacts DatalogFacts::CreateFromManifestProto(
+    const arcs::ManifestProto &manifest_proto) {
+  std::vector<ir::TagClaim> result_claims;
+  std::vector<ir::TagCheck> result_checks;
+  std::vector<ir::Edge> result_edges;
+  absl::flat_hash_map<std::string, amt::ParticleSpec> particle_specs;
+  for (const arcs::ParticleSpecProto &particle_spec_proto :
+    manifest_proto.particle_specs()) {
+    amt::ParticleSpec particle_spec =
+        amt::ParticleSpec::CreateFromProto(particle_spec_proto);
+    std::string particle_spec_name = particle_spec.name();
+    auto ins_res = particle_specs.insert({
+      std::move(particle_spec_name), std::move (particle_spec) });
```

Comment:
redundant space after `std::move`.

Also, is `std::move` even required?

Reply:
`std::move` is possibly necessary here to force the use of a move instead of a copy. It is only guaranteed that a copy will be elided if the name is no longer accessible after the copy (not no longer accessed, but no longer accessible). This is the last usage of `particle_spec` and `particle_spec_name`, but both remain accessible through the end of the loop, which has the additional statement(s) starting with the `CHECK` on the next line.

Reply:
Fixed redundant space.